src/mongo/db/curop.h

```diff
@@ -305,6 +306,10 @@ class CurOp {
         if (_dbprofile <= 0)
             return false;
 
+        if (serverGlobalParams.rateLimit > 1 && _dbprofile >= 2 && ms < serverGlobalParams.slowMS) {
+            return _prng.nextCanonicalDouble() * serverGlobalParams.rateLimit < 1.0;
```

Comment:
Well, probably it is possible to implement this logic using int only. I will think about it.
But your snippet will not do what we need: consider rateLimit=2, in this case 50% of events should be profiled.